packages/validator/src/validator.ts

```diff
@@ -67,14 +86,69 @@ export class Validator {
   private state: Status;
   private readonly controller: AbortController;
 
-  constructor(
-    opts: ValidatorOptions,
-    readonly genesis: Genesis,
-    metrics: Metrics | null = null
-  ) {
+  constructor({
+    opts,
+    genesis,
+    validatorStore,
+    slashingProtection,
+    blockProposingService,
+    attestationService,
+    syncCommitteeService,
+    config,
+    api,
+    clock,
+    chainHeaderTracker,
+    logger,
+    db,
+    metrics,
+    controller,
+  }: ValidatorModules) {
+    this.genesis = genesis;
+    this.validatorStore = validatorStore;
+    this.slashingProtection = slashingProtection;
+    this.blockProposingService = blockProposingService;
+    this.attestationService = attestationService;
+    this.syncCommitteeService = syncCommitteeService;
+    this.config = config;
+    this.api = api;
+    this.clock = clock;
+    this.chainHeaderTracker = chainHeaderTracker;
+    this.logger = logger;
+    this.controller = controller;
+    this.db = db;
+
+    if (opts.closed) {
+      this.state = Status.closed;
+    } else {
+      // "start" the validator
+      // Instantiates block and attestation services and runs them once the chain has been started.
+      this.state = Status.running;
+      this.clock.start(this.controller.signal);
+      this.chainHeaderTracker.start(this.controller.signal);
+
+      if (metrics) {
+        this.db.setMetrics(metrics.db);
+
+        this.clock.runEverySlot(() =>
+          this.fetchBeaconHealth()
+            .then((health) => metrics.beaconHealth.set(health))
+            .catch((e) => this.logger.error("Error on fetchBeaconHealth", {}, e))
+        );
+      }
+    }
+  }
+
+  get isRunning(): boolean {
+    return this.state === Status.running;
+  }
+
+  /**
+   * Initialize and start a validator client and its varied sub-component services
+   */
+  static async init(opts: ValidatorOptions, genesis: Genesis, metrics: Metrics | null = null): Promise<Validator> {
```

Comment:
You can split validator constructor logic into:
- construct all modules (sync)
- register initial validators (async)

Reply:
If a consumer of this codebase wants to start the validator without any signers, it does not need to deal with promises right?

Reply:
> You can split validator constructor logic into:

Several modules make assumptions about validators being registered already, most obvious ones are `pollPrepareBeaconProposer` and `pollBuilderValidatorRegistration`

https://github.com/ChainSafe/lodestar/blob/eaf74cda4c43a83ebc184d096e08a599420754b9/packages/validator/src/validator.ts#L198-L199

But also our duty services (`AttestationDutiesService`, `BlockDutiesService` and `SyncCommitteeDutiesService`) have a indirect assumption that the signers are already initialized because `clock.runEveryEpoch` will run immediately the first time it is called and is added in their constructor. This will result in empty duties on the first call because `getAllLocalIndices` will be empty.
https://github.com/ChainSafe/lodestar/blob/1aa656191c84e3b55281bb8f21c9f4a529282500/packages/validator/src/services/attestationDuties.ts#L146

Considering all of that, I'd say there is no easy way to split construction of modules and registering initial validators.

> If a consumer of this codebase wants to start the validator without any signers, it does not need to deal with promises right?

Yes, adding signers is the only reasoning for it to be async now but I'd argue this just gives us more flexibility in the future if there are other modules that require async initialization. Top-level await is supported everywhere, I don't really see an issue with this.